### PR TITLE
[generator] Strip the last slash from URLs after the hostname to beautify them

### PR DESCRIPTION
**Pull request by biodranik**
Also slightly reduces the size of map files

### CODE REVIEW

File: generator/osm2meta.cpp [hunk 1]
```diff
@@ -122,6 +122,23 @@ std::string MetadataTagProcessorImpl::ValidateAndFormat_operator(std::string con
...
+  // www.test.com/path => www.test.com/path
+  // www.test.com/path/ => www.test.com/path/
+  constexpr std::string_view kHttps = "https://";
+  constexpr std::string_view kHttp = "http://";
+  size_t start = 0;
+  if (strings::StartsWith(v, kHttps))
+    start = kHttps.size();
+  if (strings::StartsWith(v, kHttp))
+    start = kHttp.size();
+  auto const first = v.find('/', start);
```
[the remark this answers is not in the document]
Comment (biodranik) [reply generator/osm2meta.cpp:137]: @rtsisyk It doesn't search from the beginning. It searches from the _start_, which points _to the next symbol_ after `http://` or `https://` pfefixes.

[the remark this answers is not in the document]
Comment (vng) [reply generator/osm2meta.cpp:137]: Searching from the end can simplify this routine, at least:
```
if (!v.empty() && v.back() != '/')
  return;
```

[the remark this answers is not in the document]
Comment (vng) [reply generator/osm2meta.cpp:137]: 2. ValidateAndFormat_XXX functions better to make like void ValidateAndFormat_XXX(string & v);

[the remark this answers is not in the document]
Comment (biodranik) [reply generator/osm2meta.cpp:137]: @vng perhaps I wasn't clear. The last slash after the host name, not the last slash at the end of URL:
`www.test.com/ => www.test.com`

Changing the last non-host slash may break the URL and lead to 404:
`www.test.com/folder/ => www.test.com/folder`

[the remark this answers is not in the document]
Comment (biodranik) [reply generator/osm2meta.cpp:137]: > ValidateAndFormat_XXX functions better to make like void ValidateAndFormat_url(string & v)

@vng Agree, it requires a separate refactoring of all other cases. Does it make sense to do it separately?

File: generator/osm2meta.cpp [hunk 2]
```diff
@@ -122,6 +122,23 @@ std::string MetadataTagProcessorImpl::ValidateAndFormat_operator(std::string con
...
+  // https://www.test.com/ => https://www.test.com
+  // www.test.com/ => www.test.com
+  // www.test.com/path => www.test.com/path
+  // www.test.com/path/ => www.test.com/path/
+  constexpr std::string_view kHttps = "https://";
+  constexpr std::string_view kHttp = "http://";
+  size_t start = 0;
+  if (strings::StartsWith(v, kHttps))
+    start = kHttps.size();
+  if (strings::StartsWith(v, kHttp))
```
Comment (vng) [inline generator/osm2meta.cpp:135]: else if

File: generator/osm2meta.cpp [hunk 3]
```diff
@@ -122,6 +122,23 @@ std::string MetadataTagProcessorImpl::ValidateAndFormat_operator(std::string con
...
+    start = kHttp.size();
+  auto const first = v.find('/', start);
+  if (first == std::string::npos)
+    return v;
+  if (first == v.find_last_of('/') && first + 1 == v.size())
```
Comment (vng) [inline generator/osm2meta.cpp:140]: ```if (first + 1 == v.size())``` is enough

**Review by vng — CHANGES_REQUESTED** [review]

File: generator/osm2meta.cpp (hunk 2 above)
Comment (biodranik) [reply generator/osm2meta.cpp:135]: Good catch!

File: generator/osm2meta.cpp (hunk 3 above)
Comment (biodranik) [reply generator/osm2meta.cpp:140]: Right!

**Comment by biodranik** [comment]
@vng thanks, fixed. PTAL

**Review by vng — APPROVED** [review]

**Pull request merged**